plugins/UserCountry/Commands/AttributeHistoricalDataWithLocations.php

```diff
@@ -0,0 +1,291 @@
+        'location_city'      => LocationProvider::CITY_NAME_KEY,
+        'location_latitude'  => LocationProvider::LATITUDE_KEY,
+        'location_longitude' => LocationProvider::LONGITUDE_KEY
+    );
+
+    /**
+     * @var RawLogDao
+     */
+    protected $dao;
+
+    /**
+     * @var VisitorGeolocator
+     */
+    protected $visitorGeolocator;
+
+    /**
+     * @var int
+     */
+    private $processed = 0;
+
+    /**
+     * @var int
+     */
+    private $amountOfVisits;
+
+    /**
+     * @var Timer
+     */
+    private $timer;
+
+    /**
+     * @var int
+     */
+    private $percentStep;
+
+    /**
+     * @var int
+     */
+    private $processedPercent = 0;
+
+    public function __construct(RawLogDao $dao = null)
+    {
+        parent::__construct();
+
+        $this->dao = $dao ?: new RawLogDao();
+    }
+
+    protected function configure()
+    {
+        $this->setName('usercountry:attribute');
+
+        $this->addArgument(self::DATES_RANGE_ARGUMENT, InputArgument::REQUIRED, 'Attribute visits in this date range. Eg, 2012-01-01,2013-01-01');
+        $this->addArgument(self::PERCENT_STEP_ARGUMENT, InputArgument::OPTIONAL,
+            'How often to display the command progress. A status update will be printed after N percent of visits are processed, '
+            . 'where N is the value of this option.', self::PERCENT_STEP_ARGUMENT_DEFAULT);
+        $this->addArgument(self::PROVIDER_ARGUMENT, InputArgument::OPTIONAL, 'Provider id which should be used to attribute visits. If empty then'
+            . ' Piwik will use the currently configured provider. If no provider is configured, the default provider is used.');
+        $this->addOption(self::SEGMENT_LIMIT_OPTION, null, InputOption::VALUE_OPTIONAL, 'Number of visits to process at a time.', self::SEGMENT_LIMIT_OPTION_DEFAULT);
+    }
+
+    /**
+     * @param InputInterface $input
+     * @param OutputInterface $output
+     * @return void|int
+     */
+    protected function execute(InputInterface $input, OutputInterface $output)
+    {
+        list($from, $to) = $this->getDateRangeToAttribute($input);
+
+        $this->visitorGeolocator = $this->createGeolocator($input);
+
+        $this->percentStep = $this->getPercentStep($input);
+        $this->amountOfVisits = $this->dao->countVisitsWithDatesLimit($from, $to);
+
+        $output->writeln(
+            sprintf('Re-attribution for date range: %s to %s. %d visits to process with provider "%s".',
+                $from, $to, $this->amountOfVisits, $this->visitorGeolocator->getProvider()->getId())
+        );
+
+        $this->timer = new Timer();
+
+        $this->processSpecifiedLogsInChunks($output, $from, $to, $input->getOption(self::SEGMENT_LIMIT_OPTION));
+
+        $output->writeln("Completed. <comment>" . $this->timer->__toString() . "</comment>");
+
+        return 0;
+    }
+
+    protected function processSpecifiedLogsInChunks(OutputInterface $output, $from, $to, $segmentLimit)
+    {
```

Comment:
Maybe you can extract this into a class? Also maybe related methods like `reattributeVisitLogs `? Most likely the related methods that are used there like `reattributeVisitLogs` need to go into another class where they are public and testable. I do understand it is not as easy as you also want to output things in the middle but I'm sure there are solutions for this and if it is just an event.